### ISSUE TITLE
Required dependencies are not listed

### ISSUE THREAD

**Issue by Coder3333** [dotnet/aspnetcore#19108]
### Describe the bug
The NuGet website says that the package has no dependencies, but after installing it, when I try to run it, I get the below error message.
https://www.nuget.org/packages/dotnet-sql-cache/3.1.1

The framework 'Microsoft.NETCore.App', version '3.1.1' was not found.
  - The following frameworks were found:
      2.1.12 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      2.1.14 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      2.2.8 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
      3.1.0 at [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

You can resolve the problem by installing the specified framework and/or SDK.

The specified framework can be found at:
  - https://aka.ms/dotnet-core-applaunch?framework=Microsoft.NETCore.App&framework_version=3.1.1&arch=x64&rid=win10-x64

### To Reproduce
Install the package on a machine that does not have Microsoft.NETCore.App framework_version=3.1.1

**Comment by javiercn** [comment]
@Coder3333 thanks for contacting us.

@dotnet/aspnet-build is this something you can look into? @anurse Is this you?

**Comment by dougbu** [comment]
The NuGet Dependencies lists _packages_ required for use. Framework requirements are not specified.

In this case, you've got the 3.1.0 framework but not 3.1.1. Install the second and you'll be good to go.

**Comment by anurse** [comment]
@dougbu But the global tools shouldn't be depending upon or building using patch versions. We should only have baseline dependencies (even for framework references)...

**Comment by dougbu** [comment]
We stopped using baselines when we decided to ship everything for 3.0 and later. Guess we missed a downside. Let's discuss how broad a fix is required offline.

**Comment by Coder3333** [comment]
@dougbu, look at the dependencies for 2.1.1.  It states .NETCoreApp 2.1, where there 3.x versions do not state anything similar. 

https://www.nuget.org/packages/dotnet-sql-cache/2.1.1

.NETCoreApp 2.1
No dependencies.

**Comment by Pilchie** [comment]
@Coder3333 Framework references didn't exist in .NET Core 2, they are a new concept for 3.0, and NuGet.org is lagging in displaying them.

I do feel that we need global tools to reference the base runtime though, which is the default from the SDK.

**Comment by anurse** [comment]
> I do feel that we need global tools to reference the base runtime though, which is the default from the SDK.

That's the important part. Package references being non-baseline is OK since transitive dependencies work, but global tools need to reference the base runtime.

**Comment by Pilchie** [comment]
Package refs is not ideal either, as it causes the package version to be local, if the dependency is also in a shared framework, which means that it's then possible for it to get behind.

**Comment by Pilchie** [comment]
@dougbu - do you know if we got this fixed at some point in .NET 5.  I remember discussing it.

**Comment by dougbu** [comment]
I don't remember and see that all versions of the package after 2.1.1 lack the following, with or without an updated `targetFramework` in their .nuspec files.
``` xml
<dependencies>
  <group targetFramework=".NETCoreApp2.1" />
</dependencies>
```
I suspect errors about the installed framework come from the dotnet-sql-cache.runtimeconfig.json file e.g. from the 3.1.8 version:
``` json
"framework": {
  "name": "Microsoft.NETCore.App",
  "version": "3.1.8"
}
```
or the 5.0 preview8 version:
``` json
"framework": {
  "name": "Microsoft.NETCore.App",
  "version": "5.0.0-preview.8.20407.11"
}
```
This is fine up until RTM but not fine in servicing. We should fix this before we ship RTM if possible. Not sure exactly how&hellip;

I'm also not sure whether the `<dependencies>` element matters because the framework can be inferred from the tools/{TFM}/any folder tree. @rrelyea❔

**Comment by Pilchie** [comment]
Can we get this assigned to someone so that there is a point of contact for it?

**Comment by dougbu** [comment]
@JunTaoLuo I agree the NuGet.org / installation experience sounds annoying and may impact users of global tools that ship as packages. But, I'm not sure this affects many people. Why add this to the 5.0.0 milestone❔

@Pilchie we could assign someone to _investigate_. I suspect the main work there would be working w/ @rrelyea and his team to understand the right way forward. Is that enough for now❔

**Comment by Pilchie** [comment]
We had a meeting with @bricelam and others about this issue as it applies to `dotnet-ef`.  Also, tagging @wli3 as someone who knows a lot about .NET Global tools.

I believe that we need to do work to keep global tools referencing `Major.Minor.0` instead of `Major.Minor.Patch`, so that you can pick up a newer version of a global tool and have it work even if you haven't updated the runtime yet.

This is similar to the issues around assembly versions/package versions for things in the shared framework.  Since you've done most of the rest of that work, can you take a look at this @dougbu ?

**Comment by dougbu** [comment]
> can you take a look at this @dougbu ?

Yes though I'm unsure where `Major.Minor.0` needs to be set to get the NuGet page to reflect the correct framework requirements. In other words, changing `*.runtimeconfig.json` files in our global tools only addresses part of the reported problem. Was @rrelyea in that meeting❔

**Comment by bricelam** [comment]
We do this to dotnet-ef in 3.1:
```xml
<IsServicingBuild Condition="'$(PreReleaseVersionLabel)' == 'servicing'">true</IsServicingBuild>
```
```xml
<!--
  This keeps ef.exe targeting the default version of .NET Core for netcoreapp3.1,
  which maximizes the machines on which this tool will be compatible.
-->
<TargetLatestDotNetRuntime Condition=" '$(IsServicingBuild)' == 'true' ">false</TargetLatestDotNetRuntime>
```

**Comment by wli3** [comment]
FYI to build an runtime dependent application using SDK will give you `Major.Minor.0` by default.

**Comment by dougbu** [comment]
@wli3 do you mean `$(TargetLatestDotNetRuntime)` is no longer needed in the EF repo❔ I can't find reads of that property.

More generally, our runtimeconfig.json files definitely do _not_ currently use `Major.Minor.0` in 3.1. I also built a dotnet-sql-cache.5.0.2 package during the servicing exercise and see
``` json
{
  "runtimeOptions": {
    "tfm": "net5.0",
    "framework": {
      "name": "Microsoft.NETCore.App",
      "version": "5.0.2"
    }
  }
}
```
as it's runtimeconfig.json. Suspect we're somehow disabling the default you mentioned @wli3. What properties control this behaviour❔

**Comment by dougbu** [comment]
Separately, the other part of the problem here remains the
``` text
This package has no dependencies.
```
Shown on NuGet.org pages for these tools. What's the best way to make sure the .nuspec informs NuGet of a tool's framework dependencies❔

@bricelam the `dotnet-ef` package also neglects to mention a minimum framework needed to run the tool.

**Comment by Pilchie** [comment]
They don't have framework dependencies.  Framework dependencies is a compile time concept, meant to indicate what framework you need to be referencing in order to reference this package.  Tools are not intended to be referenced, they are meant to be executed, so they should have no dependencies in the nuspec, but have a runtimeconfig.json that indicates the runtime they will run on.

**Comment by wli3** [comment]
I cannot find `$(TargetLatestDotNetRuntime)` used anywhere outside dotnet-ef.

There should be some custom logic overriding the runtime version in aspnetcore. If you build a new template, it is not the case.

![image](https://user-images.githubusercontent.com/6993335/95239534-f1b12b00-07bf-11eb-84e8-6ec690b32189.png)

`RuntimeFrameworkVersion` property determines the number in runtimeconfig.json. But many properties would affect the end result.

**Comment by dougbu** [comment]
When `$(RuntimeFrameworkVersion)` has its default, `$(TargetLatestDotNetRuntime)` is `false`, and nothing in the repo mucks with `@(KnownFrameworkReference)` or `@(KnownFramework)` metadata (the case in dotnet/efcore), what determines the runtimeconfig.json content❔ I'm not sure what to look for in a binary log for this project.

**Comment by dougbu** [comment]
Think I found the important part of Microsoft.NET.Sdk.DefaultItems.targets:
> The general logic is that self-contained apps will target the latest patch that the SDK knows about, while framework-dependent apps will target the ".0" patch (and roll forward to the latest patch installed at runtime).

So, pretty much any project that publishes before packing could hit this but it's of particular importance in tools because those aren't packed in a way that informs customers of a shared framework version requirement. Am I on the right track❔

If yes, I _think_ we can use an approach like the `dotnet-ef` project. We do strange things in dotnet/aspnetcore like changing `@(KnownFrameworkReference)` metadata in our top-level Directory.Build.targets file but (and this may be crucial) much of that is disabled in servicing builds.

#### Bottom line
I'll try `$(TargetLatestDotNetRuntime)` out in my fake 5.0.2 branch.

**Comment by dougbu** [comment]
> I'll try `$(TargetLatestDotNetRuntime)` out in my fake 5.0.2 branch.

FYI that didn't work on its own, probably because we had an unconditional `@(FrameworkReference)` update. Trying again w/ both bits fixed&hellip;

**Comment by dougbu** [comment]
:ok: I have a fix I understand and can apply where we need it. @Pilchie should I do this for both release/3.1 and release/5.0❔ I lean toward "yes" because the original problem was reported in 3.1.x but would appreciate your confirmation.

**Comment by Pilchie** [comment]
I think yes, but we should bring these to tactics for confirmation.

**Comment by dougbu** [comment]
:ok: I'll create PRs for 3.1 and 5.0. Then, I'll send an email to Tactics.

**Comment by dougbu** [comment]
Fixes checked into release/3.1 (a8a9e00a26cc), release/5.0 (ca8cc104f4b9), and master (eb91d487c112)
